### PR TITLE
Fix EAGLE-3 context-parallel training and re-enable its tests

### PR DESCRIPTION
**Pull request by kevalmorabia97**
### What does this PR do?

Type of change: Bug fix

**EAGLE-3 context-parallel training (`--cp_size > 1`) is fixed, and its tests run again.** CP has been broken since `accelerate` 1.13, and the tests never caught it: the guard compared `Version("2.10.0a0")` against `Version("2.10.0")`, which is False on every NGC alpha torch build, so `test_llama_eagle3[cp_size=2]` has never actually run in CI.

Five fixes:

- **`main.py`** — rebuild the FSDP2 plugin accelerate requires for `cp_size > 1`. The `--fsdp full_shard --fsdp_config` launcher flags that used to supply it were dropped from `launch_train.sh`, so CP could not start at all. Also pass the CP degree to the draft model.
- **`modeling_eagle.py`** — apply the draft model's first input norm inside `layers[0]`'s own forward, where FSDP2 has actually unsharded its weights, and only stash the input embeds on the path whose pre-hook consumes them.
- **`hf_eagle.py`** — skip the dense eagle attention mask under CP (causal masking comes from `is_causal`, TTT masking from the ring-attention patch), and warn that padded positions are therefore unmasked. Also stop `(eagle_loss or 0)` replacing a `0.0` loss tensor with a plain `int`, which detached the graph.
- **`eagle_utils.py`** — key TTT-mask injection off the backward call's `grad_out` kwarg, since newer torch omits `attn_bias` on the forward call, silently disabling TTT masking.
- **`utils.py`** — CUDNN-only SDPA under CP; the `MATH` backend decomposes SDPA and breaks on DTensors. Scoped to `cp_size > 1`, since this context manager wraps every training forward and CPU has no cudnn backend.

**Drops the `speculative_decoding` 26.01 container override.** It was added when the lane ran 25.06 and spec-dec needed something *newer* — a floor. Later bumps moved the default past it, so it had silently become a ceiling holding spec-dec on a 6-month-old image.

### Testing

Ran `tests/examples/speculative_decoding` in `nvcr.io/nvidia/pytorch:26.07-py3` on 2 GPUs, reproducing the CI install steps (`pip uninstall -y nvidia-modelopt`, `pip install -e ".[hf,dev-test]"`, example requirements): **16 passed, 2 skipped** — the 2 skipped being pre-existing `--run-manual` tests. All four `test_llama_eagle3` cases pass, including both `cp_size=2` ones.

### Before your PR is "*Ready for review*"

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: N/A — the existing `cp_size=2` tests are re-enabled
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅
- Did you get Claude approval on this PR?: ❌ — not yet run

### CODE REVIEW

File: .github/workflows/example_tests.yml [hunk 1]
```diff
@@ -33,63 +33,43 @@ jobs:
...
-          - example: speculative_decoding
-            docker_image: "26.01"
+        example: [llm_distill, llm_qat, llm_sparsity, specdec_bench, speculative_decoding]
     uses: ./.github/workflows/_example_tests_runner.yml
     permissions:
       contents: read
     secrets: inherit
     with:
-      docker_image: "nvcr.io/nvidia/pytorch:${{ matrix.docker_image || '26.06' }}-py3"
+      docker_image: "nvcr.io/nvidia/pytorch:${{ matrix.docker_image || '26.07' }}-py3"
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply .github/workflows/example_tests.yml]: Fixed in f0a22fe. The lane no longer has a `docker_image` matrix entry, so the job now reads the image directly: `docker_image: "nvcr.io/nvidia/pytorch:26.07-py3"`.

File: examples/speculative_decoding/main.py [hunk 2]
```diff
@@ -186,8 +186,12 @@ def train():
...
     if training_args.cp_size > 1:
         patch_ring_attention_for_ttt()
-        # Specific patch to accelerate 1.12.0. Removable after move to 1.13.0
-        training_args.parallelism_config.sp_backend = None
+        # accelerate requires an fsdp_plugin when cp_size > 1; the --fsdp launcher flags that
+        # used to provide one were dropped from launch_train.sh.
+        if not training_args.fsdp_plugin_args:
+            training_args.fsdp = "full_shard"
+            training_args.fsdp_config = {"fsdp_version": 2}
+            training_args.fsdp_plugin_args = training_args._process_fsdp_args()
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply examples/speculative_decoding/main.py:194]: These members are inherited, not missing. `HfTrainingArguments` is built with `bases=(transformers.TrainingArguments,)` (main.py:84): `fsdp` and `fsdp_config` are dataclass fields on the base class, and `fsdp_plugin_args` is assigned in `TrainingArguments.__post_init__` (`self.fsdp_plugin_args = self._process_fsdp_args()`), so it exists on every instance. Verified on transformers 5.9 — a default instance reports `fsdp_plugin_args = None`, which is exactly the falsy value this branch tests for.

---

Both halves don't apply here.

**Transformers version:** this script runs under `examples/speculative_decoding/requirements.txt`, which pins `transformers>=5.0,<5.13`. 4.57 is the floor for the library (pyproject.toml), not for this example, and `_process_fsdp_args` / `fsdp_plugin_args` exist throughout the supported 5.x range.

**Discarding a supplied config:** the branch is guarded on `if not training_args.fsdp_plugin_args`, and a user-supplied FSDP config makes that non-empty before this line runs. Checked on transformers 5.9:

```
TrainingArguments(fsdp='full_shard', fsdp_config={'fsdp_version': 2})
  -> fsdp_plugin_args = {'fsdp_version': 2, 'reshard_after_forward': True, 'cpu_ram_efficient_loading': False, 'sync_module_states': True}
```

so the fallback is skipped and the user's settings are preserved. It only fires when nothing was configured at all, which is the case this PR fixes: `launch_train.sh` no longer passes `--fsdp`, so `cp_size > 1` could not start.

File: modelopt/torch/speculative/plugins/hf_eagle.py [hunk 3]
```diff
@@ -762,12 +762,15 @@ def forward(
...
-            # TODO: (hg) during cp training, this mask is not used. Maybe turn it off then.
             eagle_attention_mask = (
                 eagle_attn_mask_0
                 if self.eagle_mix_hidden_states or ttt_step == 0
                 else self._get_ttt_attention_mask(b, seq_length, ttt_step)
             )
+            # Under CP the dense mask is unused and fatal (plain tensor vs DTensor scores);
+            # causal masking comes from is_causal and TTT masking from the ring-attention patch.
+            if getattr(self, "eagle_cp_size", 1) > 1:
+                eagle_attention_mask = None
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply modelopt/torch/speculative/plugins/hf_eagle.py:782]: Real limitation, but pre-existing rather than introduced here, so I documented it instead of changing the attention patch.

Context parallelism never masked padding: the line this replaced carried the author's own TODO, "during cp training, this mask is not used. Maybe turn it off then." Passing the dense mask through under CP doesn't mask anything — it raises `aten.add.Tensor got mixed torch.Tensor and DTensor`, which is one of the failures this PR fixes. So the choice is between crashing and attending to pad positions, not between masked and unmasked.

Folding a padding bias into the per-ring-step TTT bias (`_get_sharded_ttt_msk`) is the real fix and is a change to the ring-attention patch, out of scope here. Added a `NOTE:` on the CP branch in 472da32 so the gap is explicit at the call site.

File: modelopt/torch/speculative/plugins/modeling_eagle.py [hunk 4]
```diff
@@ -160,7 +160,9 @@ def forward(
         # In EAGLE-3, we save input embeddings to attribute, and use it in first decoder layer by hook function
         # Also, we normalize input embeddings and hidden states before concatenating them.
         # The default input norm in first layer attn will be disabled.
-        self._input_embeds = self.layers[0].input_layernorm(inputs_embeds)
+        # Stash raw embeds and normalize in the attention pre-hook instead: FSDP2 only
+        # unshards layers[0] weights inside its own forward.
+        self._input_embeds = inputs_embeds
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply modelopt/torch/speculative/plugins/modeling_eagle.py]: Fixed in 472da32 — the stash is now `self._input_embeds = inputs_embeds if self.config.use_aux_hidden_state else None`, matching the condition that registers the consuming pre-hook (modeling_eagle.py:58-73).

Worth noting this PR already reduced the cost you describe: the previous line was `self._input_embeds = self.layers[0].input_layernorm(inputs_embeds)`, which ran a layernorm on every forward for that stash. It now keeps a reference to a tensor that already exists, and only on the path that reads it.

File: modelopt/torch/speculative/plugins/hf_eagle.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply modelopt/torch/speculative/plugins/hf_eagle.py:782]: Addressed in 2ecd662 with the second option you offered — making the limitation explicit rather than silent. The CP branch now calls `warn_rank_0` to say that padded positions are not masked and that samples should be packed or truncated to a fixed length under `cp_size > 1`.

Not a hard reject, so existing CP runs keep working, and not conditioned on the batch actually being padded: under CP the mask can be a DTensor, so inspecting it (`attention_mask.all()`) would run a collective inside a branch — the exact hazard the rest of this PR fixes. `warn_rank_0` keeps it to rank 0, and the default warning filter dedupes it to once per call site.

The padding-bias option remains the better fix and needs `_get_sharded_ttt_msk` to fold a padding mask into the per-ring-step bias; that is a change to the ring-attention patch and out of scope here.

**Review by h-guo18 — APPROVED** [review]
lgtm. The change only affect cp+eagle and is no-op for other specdec features. Thanks for the fix!
